File: bench/bun/replica.ts
// Test gRPC server in Bun — minimal walsync replica
import grpc from "@grpc/grpc-js";
import protoLoader from "@grpc/proto-loader";
import { writeFileSync, existsSync, unlinkSync, renameSync, openSync, writeSync, closeSync, statSync } from "fs";

const PROTO_PATH = new URL("../../proto/walsync.proto", import.meta.url).pathname;

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: false,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});

const walsyncProto = grpc.loadPackageDefinition(packageDefinition).walsync;

const dbPath = "/tmp/bun-replica.db";
const walPath = dbPath + "-wal";

const server = new grpc.Server();

server.addService(walsyncProto.WalSync.service, {
  ShipSnapshot(call, callback) {
    const snap = call.request;
    try {
      const dbData = snap.dbData instanceof Uint8Array ? snap.dbData : new Uint8Array(snap.dbData || []);
      const walData = snap.walData instanceof Uint8Array ? snap.walData : new Uint8Array(snap.walData || []);

      // Atomic replace
      writeFileSync(dbPath + ".tmp", dbData);
      if (existsSync(walPath)) { try { unlinkSync(walPath); } catch {} }
      try { unlinkSync(dbPath + "-shm"); } catch {}
      renameSync(dbPath + ".tmp", dbPath);
      if (walData.length > 0) {
        writeFileSync(walPath, walData);
      }
      console.log(`snapshot received: ${dbData.length} bytes db, ${walData.length} bytes wal`);
      callback(null, { ok: true, error: "", applied_offset: 0 });
    } catch (e) {
      console.error("snapshot error:", e);
      callback(null, { ok: false, error: String(e), applied_offset: 0 });
    }
  },

  ShipWal(call, callback) {
    const chunk = call.request;
    if (!chunk.data || chunk.data.length === 0) {
      callback(null, { ok: true, error: "", applied_offset: 0 });
      return;
    }
    try {
      const offset = Number(chunk.offset);
      const data = chunk.data instanceof Uint8Array ? chunk.data : new Uint8Array(chunk.data);
      const fd = openSync(walPath, existsSync(walPath) ? "r+" : "w");
      writeSync(fd, data, 0, data.length, offset);
      closeSync(fd);
      console.log(`WAL received: ${data.length} bytes at offset ${offset}`);
      callback(null, { ok: true, error: "", applied_offset: offset + data.length });
    } catch (e) {
      console.error("wal error:", e);
      callback(null, { ok: false, error: String(e), applied_offset: 0 });
    }
  },

  Health(call, callback) {
    const dbSize = existsSync(dbPath) ? statSync(dbPath).size : 0;
    const walSize = existsSync(walPath) ? statSync(walPath).size : 0;
    callback(null, { ok: true, db_size: dbSize, wal_size: walSize });
  },
});

server.bindAsync("127.0.0.1:9192", grpc.ServerCredentials.createInsecure(), (err, port) => {
  if (err) {
    console.error("failed to bind:", err);
    process.exit(1);
  }
  console.log(`Bun gRPC replica listening on 127.0.0.1:${port}`);
});

File: src/replica.rs
use crate::proto::wal_sync_server::{WalSync, WalSyncServer};
use crate::proto::{Ack, Empty, HealthResponse, Snapshot, WalChunk};
use crate::util;
use log::{error, info};
use std::pin::Pin;
use tonic::codec::CompressionEncoding;
use tonic::transport::Server;
use tonic::{Request, Response, Status};

pub struct ReplicaServer {
    db_path: String,
    wal_path: String,
}

impl ReplicaServer {
    pub fn new(db_path: &str) -> Self {
        Self {
            db_path: db_path.to_string(),
            wal_path: format!("{}-wal", db_path),
        }
    }
}

#[tonic::async_trait]
impl WalSync for ReplicaServer {
    async fn ship_snapshot(
        &self,
        request: Request<Snapshot>,
    ) -> Result<Response<Ack>, Status> {
        let snap = request.into_inner();

        // Atomic replace: write to .tmp, remove stale WAL/SHM, rename
        if let Err(e) = util::atomic_replace_db(&self.db_path, &snap.db_data) {
            error!("snapshot write error: {}", e);
            return Ok(Response::new(Ack {
                ok: false,
                error: e.to_string(),
                applied_offset: 0,
            }));
        }

        // Write WAL if provided
        if !snap.wal_data.is_empty() {
            if let Err(e) = std::fs::write(&self.wal_path, &snap.wal_data) {
                error!("snapshot wal write error: {}", e);
                return Ok(Response::new(Ack {
                    ok: false,
                    error: e.to_string(),
                    applied_offset: 0,
                }));
            }
        }

        info!(
            "snapshot received: {} bytes db, {} bytes wal",
            snap.db_data.len(),
            snap.wal_data.len()
        );
        Ok(Response::new(Ack {
            ok: true,
            error: String::new(),
            applied_offset: 0,
        }))
    }

    async fn ship_wal(
        &self,
        request: Request<WalChunk>,
    ) -> Result<Response<Ack>, Status> {
        let chunk = request.into_inner();

        if chunk.data.is_empty() {
            return Ok(Response::new(Ack {
                ok: true,
                error: String::new(),
                applied_offset: 0,
            }));
        }

        let offset = chunk.offset;

        if let Err(e) = util::write_file_at(&self.wal_path, offset, &chunk.data) {
            error!("wal write error: {}", e);
            return Ok(Response::new(Ack {
                ok: false,
                error: e.to_string(),
                applied_offset: 0,
            }));
        }

        info!(
            "WAL received: {} bytes at offset {}",
            chunk.data.len(),
            offset
        );
        Ok(Response::new(Ack {
            ok: true,
            error: String::new(),
            applied_offset: offset + chunk.data.len() as i64,
        }))
    }

    async fn health(
        &self,
        _request: Request<Empty>,
    ) -> Result<Response<HealthResponse>, Status> {
        Ok(Response::new(HealthResponse {
            ok: true,
            db_size: util::file_size(&self.db_path),
            wal_size: util::file_size(&self.wal_path),
        }))
    }
}

pub async fn run_replica(db_path: &str, listen: &str) {
    info!("walsync replica starting | db={} | listen={}", db_path, listen);

    let bind_addr = if listen.starts_with(':') {
        format!("0.0.0.0{}", listen)
    } else {
        listen.to_string()
    };

    let addr: std::net::SocketAddr = match bind_addr.parse() {
        Ok(a) => a,
        Err(e) => {
            error!("failed to parse listen address {}: {}", bind_addr, e);
            std::process::exit(1);
        }
    };

    let server = ReplicaServer::new(db_path);

    info!("replica listening on {} (gRPC)", listen);

    if let Err(e) = Server::builder()
        .add_service(
            WalSyncServer::new(server)
                .accept_compressed(CompressionEncoding::Gzip)
                .send_compressed(CompressionEncoding::Gzip),
        )
        .serve(addr)
        .await
    {
        error!("server error: {}", e);
        std::process::exit(1);
    }
}
